### PR TITLE
feat: turbo

### PR DESCRIPTION
**Pull request by sdelamo**
This simplifies using [Turbo](https://turbo.hotwired.dev) with Micronaut applications.

Close: https://github.com/micronaut-projects/micronaut-views/issues/156

### CODE REVIEW

**Comment by sdelamo** [comment]
@timyates I am missing the docs but If you can check the code it would be great.

**Comment by timyates** [comment]
I assume turbo works along-side existing templating libs?

**Comment by timyates** [comment]
All the sonar bugs appear to be false positives (which is a bit annoying) https://sonarcloud.io/project/issues?id=micronaut-projects_micronaut-views&pullRequest=365&resolved=false&types=BUG

**Comment by timyates** [comment]
I've confirmed the test failures locally on this branch... jdk 17 gives us

```
 Exception in thread "io-executor-thread-1" groovy.lang.GroovyRuntimeException
        at org.codehaus.groovy.vmplugin.v9.Java9.of(Java9.java:166)
        at org.codehaus.groovy.vmplugin.v9.Java9.getInvokeSpecialHandle(Java9.java:179)
        at org.codehaus.groovy.runtime.ConversionHandler.lambda$invoke$0(ConversionHandler.java:97)
        at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
        at org.codehaus.groovy.runtime.ConversionHandler.invoke(ConversionHandler.java:96)
        at jdk.proxy3/jdk.proxy3.$Proxy53.writeTo(Unknown Source)
        at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$encodeHttpResponse$6(RoutingInBoundHandler.java:946)
        at io.micronaut.scheduling.instrument.InvocationInstrumenterWrappedRunnable.run(InvocationInstrumenterWrappedRunnable.java:47)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
  Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.codehaus.groovy.vmplugin.v9.Java9.of(Java9.java:160)
        ... 10 more
  Caused by: java.lang.IllegalAccessException: module jdk.proxy3 does not open jdk.proxy3 to unnamed module @63c12fb0
        at java.base/java.lang.invoke.MethodHandles.privateLookupIn(MethodHandles.java:259)
        ... 15 more
```

**Comment by graemerocher** [comment]
Perhaps because Groovy was downgraded for some reason? https://github.com/micronaut-projects/micronaut-views/pull/365/files#diff-3d103fc7c312a3e136f88e81cef592424b8af2464c468116545c4d22d6edcf19R5

**Comment by timyates** [comment]
Don't think so, [I'm seeing the same locally](https://ge.micronaut.io/s/pezjinky5rw2e/dependencies?focusedDependency=WzAsOCwyNzAsWzAsOCxbMzc1LDI3MF1dXQ&toggled=W1swXSxbMCw4XSxbMCw4LFszNzVdXV0) with 3.0.10

**Comment by timyates** [comment]
This is the line it's failing on 🤔

https://github.com/micronaut-projects/micronaut-views/blob/734651c9b1357b0140cefd9a11571b5217bd1936/views-core/src/test/groovy/io/micronaut/views/turbo/TurboStreamSpec.groovy#L166

File: gradle.properties [hunk 1]
```diff
@@ -2,7 +2,7 @@ projectVersion=3.2.0-SNAPSHOT
 projectGroup=io.micronaut.views
 micronautVersion=3.3.4
 micronautTestVersion=3.0.5
-groovyVersion=3.0.10
+groovyVersion=3.0.9
```
Comment (timyates) [inline gradle.properties:5]: Should be 3.0.10

**Review by timyates — CHANGES_REQUESTED** [review]

**Comment by sdelamo** [comment]
@graemerocher @timyates this refactor fixed the Groovy 17 bug: 

https://github.com/micronaut-projects/micronaut-views/pull/365/commits/1eb30ba751130cea911f51c18b66fddcf9682541

File: gradle.properties (hunk 1 above)
Comment (sdelamo) [reply gradle.properties]: Resolved via https://github.com/micronaut-projects/micronaut-views/pull/365/commits/5e209423a733b6e8413f053947f54a7d1db2d72e

**Review by timyates — APPROVED** [review]

**Comment by sdelamo** [comment]
> I wonder if it was splitting the test, returning a new Writable or both that fixes it 🤔

I split the test but still failed. It was not until I remove the lambda that worked.

**Comment by sdelamo** [comment]
I wrote a Micronaut implementation of [Hotwire Rails demo Chat application](https://github.com/hotwired/hotwire-rails-demo-chat) which is shown in the [Hotwire intro video](https://hotwired.dev). 

https://github.com/sdelamo/hotwire-micronaut-demo-chat

**Comment by sdelamo** [comment]
@timyates can you review this again?. I have added docs. I think it is ready to be merged.

**Comment by timyates** [comment]
👀

File: views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+
+    @NonNull
+    private Optional<String> getTargetCssQuerySelectorHtmlAttribute() {
+        return getTargetCssQuerySelector()
+                .map(cssSelector -> SPACE + htmlAttribute(TURBO_STREAM_ATTRIBUTE_TARGETS, cssSelector));
+    }
+
+    @NonNull
+    private String htmlAttribute(@NonNull String key, @NonNull String value) {
+        return String.join(EQUALS, key, DOUBLE_QUOTE + value + DOUBLE_QUOTE);
```
[the remark this answers is not in the document]
Comment (timyates) [reply views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java:181]: I cannot get a script injected by manipulating it in a demo app, so we may be ok as I believe turbo does cleaning and checks

**Review by timyates — APPROVED** [review]
Very nice!  And the Chat app should be a guide 😻

File: views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sdelamo) [reply views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java:181]: not sure I understand the concern. This is private method to create an html attribute such as: 

htmlAttribute("target", "message_1") == 'target="message_1"'

[the remark this answers is not in the document]
Comment (timyates) [reply views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java:181]: It was if someone uses a user supplied input for the dom id:

```
TurboStream turboStream = TurboStream.builder()
    .action(TurboStreamAction.APPEND)
    .targetDomId(domId)
    .template("Content to append to container designated with the dom_id.")
    .build();
Optional<Writable> writable = turboStream.render();
```
If `domId` is `id"><template><script>alert(1);</script></template><p id="`

Then the generated HTML is:

```
<turbo-stream action="append" target="id"><template><script>alert(1);</script></template><p id="">
    <template>
        Content to append to container designated with the dom_id.
    </template>
</turbo-stream>
```

But as I said, I cannot get this to show an alert, so it's probably nothing

[the remark this answers is not in the document]
Comment (timyates) [reply views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java:181]: However, escaping or removing the `"` (if present) would shut the door (if there is a door)

**Review by timyates — APPROVED** [review]
Added a check for `"` to the tests

Nice 😄

File: views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+            Optional<String> turboFrameOptional = request.getHeaders().get(TurboHttpHeaders.TURBO_FRAME, String.class);
+            if (!route.hasAnnotation(TurboView.class) ||
+                    (!TurboMediaType.acceptsTurboStream(request) && !turboFrameOptional.isPresent() && route.hasAnnotation(View.class))) {
+                return Optional.empty();
+            }
+            TurboStream.Builder builder = TurboStream.builder();
+            route.getValue(TurboView.class, String.class).ifPresent(builder::templateView);
+            route.getValue(TurboView.class, MEMBER_ACTION, TurboStreamAction.class).ifPresent(builder::action);
+            route.getValue(TurboView.class, MEMBER_TARGET_DOM_ID, String.class).ifPresent(builder::targetDomId);
+            route.getValue(TurboView.class, MEMBER_TARGET_CSS_QUERY_SELECTOR, String.class).ifPresent(builder::targetCssQuerySelector);
```
Comment (graemerocher) [inline views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java:474]: using `stringValue(..)` and `enumValue(..)` is more performant

File: views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+                    AnnotationMetadata.class);
+            if (!routeMatch.isPresent()) {
+                return Optional.empty();
+            }
+            return of(request, routeMatch.get());
+        }
+
+        private static Optional<TurboStream.Builder> of(@NonNull HttpRequest<?> request,
+                                                        @NonNull AnnotationMetadata route) {
+            Optional<String> turboFrameOptional = request.getHeaders().get(TurboHttpHeaders.TURBO_FRAME, String.class);
```
Comment (graemerocher) [inline views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java:465]: using stringValue(..) is more performant

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: views-core/src/main/java/io/micronaut/views/ViewsFilter.java [hunk 5]
```diff
@@ -123,4 +168,18 @@ protected MediaType resolveMediaType(@NonNull HttpResponse<?> response, @Nullabl
...
+        Optional<TurboStream.Builder> turboStreamBuilderOptional = TurboStream.Builder.of(request, response);
+        if (turboStreamBuilderOptional.isPresent()) {
+            return Optional.of(turboStreamBuilderOptional.get()
+                    .templateModel(body));
+        } else if (body instanceof TurboStream.Builder) {
+            return Optional.of((TurboStream.Builder) body);
+        }
+        return Optional.empty();
+    }
 }
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply views-core/src/main/java/io/micronaut/views/ViewsFilter.java]: https://github.com/micronaut-projects/micronaut-views/pull/365/commits/94d053388239e12020b8a05c36e721edb804de18

File: views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java (hunk 3 above)
Comment (sdelamo) [reply views-core/src/main/java/io/micronaut/views/turbo/TurboStream.java]: `enumValue` does not work. I think we have a bug. I am going to report it in core.

**Pull request merged**